src/main/java/io/tiledb/java/api/Group.java

```diff
@@ -24,12 +24,403 @@
+
+    NativeArray result = new NativeArray(ctx, nativeType, value, (int) valueLength);
+
+    tiledb.delete_uintp(value_num);
+    tiledb.delete_uintp(key_len);
+    tiledb.delete_charpp(key);
+    tiledb.delete_tiledb_datatype_tp(value_type);
+
+    return new Pair<String, NativeArray>(keyString, result);
+  }
+
+  /**
+   * Checks if the key is present in the group metadata. The array must be opened in READ mode,
+   * otherwise the function will error out.
+   *
+   * @param key a key to retrieve from the metadata key-value
+   * @return true if the key is present in the metadata, false if it is not
+   * @throws TileDBError A TileDB exception
+   */
+  public Boolean hasMetadataKey(String key) throws TileDBError {
+    if (!isOpen()) throw new TileDBError("Group with URI: " + uri + " is closed");
+    SWIGTYPE_p_tiledb_datatype_t value_type = tiledb.new_tiledb_datatype_tp();
+    SWIGTYPE_p_int has_key = tiledb.new_intp();
+
+    ctx.handleError(
+        tiledb.tiledb_group_has_metadata_key(ctx.getCtxp(), getGroupp(), key, value_type, has_key));
+
+    Boolean result = tiledb.intp_value(has_key) > 0;
+
+    tiledb.delete_intp(has_key);
+    tiledb.delete_tiledb_datatype_tp(value_type);
+
+    return result;
+  }
+
+  /**
+   * Check if the group is open.
+   *
+   * @return True if the group is open.
+   * @throws TileDBError
+   */
+  public boolean isOpen() throws TileDBError {
+    boolean isOpen;
+    SWIGTYPE_p_int ret = tiledb.new_intp();
+    try {
+      ctx.handleError(tiledb.tiledb_group_is_open(ctx.getCtxp(), getGroupp(), ret));
+      isOpen = tiledb.intp_value(ret) != 0;
+    } finally {
+      tiledb.delete_intp(ret);
+    }
+    return isOpen;
+  }
+
+  /**
+   * * It puts a metadata key-value item to an open group. The group must * be opened in WRITE mode,
+   * otherwise the function will error out.
+   *
+   * @param key The key of the metadata item to be added.
+   * @param value The metadata value.
+   * @throws TileDBError
+   */
+  public void putMetadata(String key, NativeArray value) throws TileDBError {
+    if (!isOpen()) throw new TileDBError("Group with URI: " + uri + " is closed");
+    ctx.handleError(
+        tiledb.tiledb_group_put_metadata(
+            ctx.getCtxp(),
+            getGroupp(),
+            key,
+            value.getNativeType().toSwigEnum(),
+            value.getSize(),
+            value.toVoidPointer()));
+  }
+
+  /** Close resources */
+  public void close() {
+    if (groupp != null && grouppp != null) {
+      tiledb.tiledb_group_close(ctx.getCtxp(), groupp);
+      tiledb.tiledb_group_free(grouppp);
+      grouppp = null;
+      groupp = null;
+    }
+  }
+
+  /**
+   * This method is responsible for closing the group object without deleting its pointers. It is
+   * useful for cases were we need to reopen a group. The "close" method above deletes a group
+   * completely.
+   *
+   * @throws TileDBError
+   */
+  public void sleep() throws TileDBError {
```

Comment:
This method is equivalent to the close() method in C++. However all objects in java need to implement the autoclosable interface which contains a "close()" method in which all pointers and the object itself is deleted. Close() in c++ has the meaning of sleep. This is the reasoning behind the naming here.

Reply:
I don't think `sleep` is quite right here. `detach` or `release` or `disengage` seems closer to the meaning we need here.

Reply:
Actually I think we should:
- hide `.open` and take the mode in the constructor as in the earlier comment (so `.open` would become private only)
- implement `reopen` instead, similar to the `Array` API, and then make this call to `tiledb_group_close` inside of `reopen`
eg:
```
void reopen(mode) throws TileDBError {
  tiledb_group_close(self)
  self._private_open(mode);
}
```

Reply:
Done!